text/3714-macro-fragment-fields.md

```diff
@@ -0,0 +1,182 @@
+
+# Summary
+[summary]: #summary
+
+Add a syntax and mechanism for macros to access "fields" of high-level fragment
+specifiers that they've matched, to let macros use the Rust parser for
+robustness and future compatibility, while still extracting pieces of the
+matched syntax.
+
+# Motivation
+[motivation]: #motivation
+
+The macros-by-example system is powerful, but sometimes difficult to work with.
+In particular, parsing complex parts of Rust syntax often requires carefully
+recreating large chunks of the Rust grammar, in order to parse out the desired
+pieces. Missing or incorrectly handling any portion of the syntax can result in
+not accepting the same syntax Rust does; this includes future extensions to
+Rust syntax that the macro was not yet aware of. Higher-level fragment
+specifiers are more robust for these cases, but don't allow extracting
+individual pieces of the matched syntax.
+
+This RFC introduces a mechanism to use high-level fragment specifiers while
+still extracting individual pieces of the matched syntax.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+When writing macros by example, and using certain high-level fragment
+specifiers, you can use the syntax `${matched_name.field_name}` to extract
+specific "fields" of the matched syntax. This allows you to use the Rust parser
+for those high-level fragments, rather than having to recreate parts of the
+Rust grammar in order to extract the specific pieces you want. Fields evaluate
+to pieces of Rust syntax, suitable for substitution into the program or passing
+to other macros for further processing.
+
+For example, the fragment `:adt` parses any abstract data type supported by
+Rust: struct, union, or enum. Given a match `$t:adt`, you can obtain the name
+of the matched type with `${t.name}`:
+
+```rust
+macro_rules! get_name {
+    ($t:adt) => { println!("{}", stringify!(${t.name})); }
+}
+
+fn main() {
+    let n1 = get_name!(struct S { field: u32 });
+    let n2 = get_name!(enum E { V1, V2 = 42, V3(u8) });
+    let n3 = get_name!(union U { u: u32, f: f32 });
+    println!("{n3}{n1}{n2}"); // prints "USE"
+}
+```
+
+An attempt to access a field that doesn't exist will produce a compilation
+error on the macro definition, whether or not the specific macro rule gets
+invoked.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Fragment fields may be used in a macro transcriber anywhere the corresponding
+fragment name could be used.
+
+Fragment fields typically follow the same rules for repetition handling as the
+corresponding fragment (e.g. being used at the same level/kind of repetition).
+However, fragment fields that contain multiple items require one additional
+level of repetition; see the `param` field of `:fn`, below.
+
+This RFC introduces the following new fragment specifiers, with specified fields:
```

Comment:
For disambiguity here, it might be nice to specify the specific AST node as in the [reference](https://doc.rust-lang.org/reference/items.html) - e.g. `:fn` is specifically a <code>Visibility<sup>?</sup> Function</code>, and I can't tell whether `:adt` is a <code>(Struct | Enumeration | Union)</code> or a <code>Visibility<sup>?</sup> (Struct | Enumeration | Union)</code>

Reply:
I expect visibility to match the `:vis` matcher, so a visibility can be empty but it isn't optional. so `Visibility Function` rather than <code>Visibility<sup>?</sup> Function</code>

Reply:
I've added `vis` for `:adt`.

To the more general point: I agree in principle that documentation for fragment specifiers may want to reference specific AST nodes, but I think the two introduced here should now be unambiguous.

Reply:
@programmerjake yes, for `:vis` - I was just going by the convention the reference page I linked uses, which had <code>VisItem: Visibility<sup>?</sup> (Module | ExternCrate | UseDeclaration | ...)</code>. But in any case.